Cilsil/Cil/Parsers/LdlocParser.cs

```diff
@@ -72,6 +80,14 @@ internal class LdlocParser : InstructionParser
                                                                variableType,
                                                                variableExpression));
             }
+            else if (value != null && value is BinopExpression && type is Tptr)
+            {
+                // For if/else/loop branching, we add the binop expression in condition.
```

Comment:
This modification seems strange. You are not only outright converting a Tptr into a completely different type (IBool), but also you are foregoing the production of an actual SIL instruction while still pushing something onto the stack. Additionally, you are doing this in a fairly common scenario (basically any binop that involves an object type). While this approach might alleviate issues in certain scenarios, I suspect it can create issues in others, unless you have some kind of theoretical guarantee on the correctness of this in all possible situations where this case can arise. 

Deviations from the standard handling should have very precise conditions under which they are triggered. For example, in handling isinst I specifically look for the ">" as well as the leftExpressionType being the product of an isinst translation. This helps avoid issues in which the special handling is triggered in cases that do not merit it. 

![image](https://user-images.githubusercontent.com/38902361/122342185-281b7a00-cef9-11eb-952d-df2fd6b46384.png)

Reply:
Done. In this [commit](https://github.com/microsoft/infersharp/pull/57/commits/3d4687173eb44a74fb7c7cdbdbeb22f1a51b7078)

Reply:
Before we used to load the condition as an extra instruction and only put its identifier name in the PRUNE, which would cause Infer confused and ignore, for example, null checks. Like the example shown below

![image](https://user-images.githubusercontent.com/63893533/123143905-ed8d7200-d40f-11eb-8bf2-2203eea38fe4.png)

Which causes infer thinks this "ResourceLeakOK" has a resource leak because Infer didn't recognize the null check and therefore didn't know node 85 connects to exit node is Ok. 

In addition, as you can see the cfg with our previous translation logic (the cfg on the right hand side of the diff), we represent n$4 as Ibool, which is actually "n$3 == null". What I am doing in your screenshot is just to define "n$3 == null"'s type is Ibool.

Reply:
Hmmm -- is the righthand side (with n$4) the latest version? It shouldn't be interpretable by Infer, as n$4 is a load from %1, which itself does not exist in the context.